### PR TITLE
Detect Running App on Start

### PR DESCRIPTION
**Pull request by makluganteng**
Work in progress 
Pull Request related to #25

### CODE REVIEW

File: src/main/madara-app.ts [hunk 1]
```diff
@@ -190,7 +190,93 @@ export async function startApp(window: BrowserWindow, appId: string) {
...
+    // );
+
+    const containerName = getContainerName(appConfig.containers[0], appConfig);
+
+    const runningContainers = await docker.listContainers();
+    // const isAppRunning = runningContainers.some((container) =>
+    //   containerNames.includes(container.Names[0].replace(/^\//, ''))
+    // );
+    const isAppRunning = runningContainers.some(
+      (container) => container.Names[0].replace(/^\//, '') === containerName
```
Comment (apoorvsadana) [inline src/main/madara-app.ts]: what is this replace for? can we add a comment as well for how the name are before the replace?

File: src/main/madara-app.ts [hunk 2]
```diff
@@ -190,7 +190,93 @@ export async function startApp(window: BrowserWindow, appId: string) {
...
+    //   getContainerName(container, appConfig)
+    // );
+
+    const containerName = getContainerName(appConfig.containers[0], appConfig);
+
+    const runningContainers = await docker.listContainers();
+    // const isAppRunning = runningContainers.some((container) =>
+    //   containerNames.includes(container.Names[0].replace(/^\//, ''))
+    // );
+    const isAppRunning = runningContainers.some(
```
Comment (apoorvsadana) [inline src/main/madara-app.ts]: this code shouldn't be on startApp. it should actually be done whenever Apps.tsx loads. so you can have a function inside madara-app.ts that fetches all running apps and the `Apps.tsx` component in react can call this function via the IPC

File: src/main/madara-app.ts [hunk 3]
```diff
@@ -190,7 +190,93 @@ export async function startApp(window: BrowserWindow, appId: string) {
   if (appConfig.appType === 'binary') {
     await startAppBinary(window, appConfig);
   } else if (appConfig.appType === 'docker') {
-    await startAppDocker(window, appConfig);
```
Comment (apoorvsadana) [inline src/main/madara-app.ts]: we shouldn't be commenting this out, i guess you did it for testing but still adding the comment so we don't forget

File: src/main/madara-app.ts [hunk 4]
```diff
@@ -252,3 +252,58 @@ export async function getAppSettings(appId: string) {
...
+
+export async function fetchAllRunningApps(window: BrowserWindow) {
+  const containers = await docker.listContainers();
+  const runningApps: { [appId: string]: boolean } = {};
+
+  containers.forEach((container) => {
+    const containerName = container.Names[0].substring(1); // remove the leading "/"
+    const parts = containerName.split('-');
+    const partialAppId = parts.slice(1, 5).join('-');
+    runningApps[partialAppId] = true;
```
Comment (apoorvsadana) [inline src/main/madara-app.ts]: we can make runningApps an array directly instead of doing `.keys` eventually

File: src/main/madara-app.ts [hunk 5]
```diff
@@ -252,3 +252,58 @@ export async function getAppSettings(appId: string) {
...
+  });
+
+  const partialAppIds = Object.keys(runningApps);
+
+  const appsFromConfigRunning = APPS_CONFIG.apps.filter((app) =>
+    partialAppIds.some((partialAppId) => app.id.startsWith(partialAppId))
```
Comment (apoorvsadana) [inline src/main/madara-app.ts]: why can't we directly equate (instead of `startsWith`)

File: src/main/madara-app.ts [hunk 6]
```diff
@@ -252,3 +252,58 @@ export async function getAppSettings(appId: string) {
...
+  apps.map(async (app) => {
+    const containers = await docker.listContainers({
+      filters: {
+        name: [`${app.id}-*`],
+      },
+    });
+    await Promise.all(
+      containers.map(async (container) => {
+        const containerApp = await docker.getContainer(container.Id);
+        containerApp.attach(
```
Comment (apoorvsadana) [inline src/main/madara-app.ts]: this code of sending logs is duplicated, can we move it to a common function to remove redundant code?

File: src/main/madara-app.ts (hunk 5 above)
Comment (makluganteng) [reply src/main/madara-app.ts]: This is cause by the container read from the docker is using 4/5 of the id from the one in the appsconfig's id

Comment (makluganteng) [reply src/main/madara-app.ts]: i can use include to simplify the logic actually

Comment (apoorvsadana) [reply src/main/madara-app.ts]: What exactly do you mean by 4/5? Also, it's usually a good practice to for the person who created the comment to resolve the issue. This way, we ensure that we only resolve once both of us agree on the solution :)

File: src/main/madara-app.ts (hunk 6 above)
Comment (apoorvsadana) [reply src/main/madara-app.ts]: I don't think this is fixed?

File: src/main/madara-app.ts (hunk 5 above)
Comment (makluganteng) [reply src/main/madara-app.ts]: so example like the container name is 'tsukuyomi-6430cd1b-097d-46be-a44d-aa7631433910-apibara-starknet
' and the id to that app is 6430cd1b-097d-46be-a44d-aa7631433910 thats why i did a slice on the top to join 1 and 5. Maybe any idea on a better way to tackle this ?

File: src/main/madara-app.ts (hunk 6 above)
Comment (makluganteng) [reply src/main/madara-app.ts]: My bad, Which is the code redundant to ?

File: src/main/madara-app.ts (hunk 5 above)
Comment (apoorvsadana) [reply src/main/madara-app.ts]: Ahh ok, that's not so neat. Instead lets do `containerName.includes(app.id)`

File: src/main/madara-app.ts (hunk 6 above)
Comment (apoorvsadana) [reply src/main/madara-app.ts]: this one

```
container.attach(
    { stream: true, stdout: true, stderr: true },
    (err, stream) => {
      if (!stream) return;
      stream.on('data', (data) => {
        // removing the first 8 bytes as they have information about the stream
        const buffer = Buffer.from(data).subarray(8, data.length);
        const event: AppAppendLogs = {
          appId: appConfig.id,
          containerName: containerConfig.name as string,
          logs: buffer.toString(),
        };
        window.webContents.send('app-logs', event);
      });
    }
  );
```

Comment (makluganteng) [reply src/main/madara-app.ts]: i fixed the redundant in the latest commit

File: src/main/madara-app.ts (hunk 5 above)
Comment (makluganteng) [reply src/main/madara-app.ts]: Okey noted

Comment (makluganteng) [reply src/main/madara-app.ts]: did u mean like this 
`const appsFromConfigRunning = APPS_CONFIG.apps.filter(
    (app) =>
      app.appType === 'docker' &&
      runningApps.some((partialAppId) => app.id.includes(partialAppId))
  );`

**Pull request merged**